cmd/api/src/model/tokenstring.go

```diff
@@ -0,0 +1,169 @@
+package model
+
+import (
+	"crypto/rand"
+	"database/sql/driver"
+	"encoding/base64"
+	"errors"
+	"fmt"
+	"hash/crc32"
+	"io"
+	"math/big"
+	"strings"
+)
+
+func generateTokenValue(prng io.Reader) (string, error) {
+	var val strings.Builder
+	big_range := big.NewInt(62) // a-zA-Z0-9 gives a range of 62 chars
+	val.Grow(64)
+	for i := 0; i < 64; i++ {
+		if n, err := rand.Int(prng, big_range); err != nil {
+			return "", fmt.Errorf("error getting random value: %w", err)
+		} else {
+			if _, err := val.WriteString(n.Text(62)); err != nil {
+				return "", fmt.Errorf("error appending character to string: %w", err)
+			}
+		}
+	}
+	return val.String(), nil
+}
+
+type TokenString struct {
+	Prefix    string
+	value     string
+	cksum     uint32
+	is_legacy bool
+}
+
+func GenerateTokenString(prefix string) (TokenString, error) {
+	if len(prefix) <= 0 {
+		return TokenString{}, fmt.Errorf("prefix must not be empty")
+	}
+	if val, err := generateTokenValue(rand.Reader); err != nil {
+		return TokenString{}, fmt.Errorf("error generating token string value: %w", err)
+	} else {
+		return CreateTokenStringWithValue(prefix, val)
+	}
+}
+
+func CreateTokenStringWithValue(prefix, value string) (TokenString, error) {
+	if len(prefix) <= 0 {
+		return TokenString{}, fmt.Errorf("prefix must not be empty")
+	}
+	if len(value) != 64 {
+		return TokenString{}, fmt.Errorf("value must be of length 64")
+	}
+	new := TokenString{Prefix: strings.ToUpper(prefix), value: value}
+	new.cksum = crc32.ChecksumIEEE([]byte(new.Prefix + new.value))
+	return new, nil
+}
+
+func formatChecksum(cksum uint32) string {
+	return strings.ReplaceAll(fmt.Sprintf("%6s", big.NewInt(int64(cksum)).Text(62)), " ", "0")
+}
+
+// This method isn't really necessary anymore, but leaving it in case
+// we want to modify what part of a token is hashed in the future
+func (s TokenString) DigestableValue() ([]byte, error) {
+	if s.value == "" {
+		return []byte{}, errors.New("token value is not set")
+	} else {
+		return []byte(s.String()), nil
+	}
+}
+
+func (s TokenString) String() string {
+	if s.value == "" {
+		return ""
+	} else if s.is_legacy {
+		return s.value
+	}
+	return fmt.Sprintf("%s_%s%s", s.Prefix, s.value, formatChecksum(s.cksum))
+}
+
+func (s TokenString) MarshalText() ([]byte, error) {
+	return []byte(s.String()), nil
+}
+
+func (s *TokenString) loadFromString(src string) error {
```

Comment:
I agree and you should consider removing this method.

It doesn't appear that the two other pointer receiver methods that depend on this are used outside of testing. Structs with a predictably small memory footprint should prefer parsing factory functions instead of using mutating functions. That said, my personal bias is to generally avoid mutator functions when I can avoid it so I'm inclined to file this and my related comments under a non-blocking nit :wink:

Reply:
I too don't particularly like the pattern, however as noted below those dependent functions are to fulfill the proper interfaces to support loading from JSON and the DB.